drivers/lora/rylrxxx.c

```diff
@@ -0,0 +1,632 @@
+	const struct rylr_config *config = dev->config;
+	struct uart_config dt_uart_config;
+	const struct modem_cmd_handler_config cmd_handler_config = {
+		.match_buf = data->cmd_match_buf,
+		.match_buf_len = CONFIG_LORA_RYLRXX_CMD_BUF_SIZE,
+		.buf_pool = data->buf_pool,
+		.alloc_timeout = RYLR_ALLOC_TIMEOUT,
+		.eol = RYLR_CMD_TERMINATOR,
+		.user_data = data,
+		.response_cmds = response_cmds,
+		.response_cmds_len = ARRAY_SIZE(response_cmds),
+		.unsol_cmds = unsolicited_cmds,
+		.unsol_cmds_len = ARRAY_SIZE(unsolicited_cmds),
+	};
+	data->pending_async_flags = 0;
+	data->dev = dev;
+	err = modem_cmd_handler_init(&data->mctx.cmd_handler, &data->cmd_handler_data,
+				     &cmd_handler_config);
+	if (err != 0) {
+		LOG_ERR("modem cmd handler init err: %d", err);
+		return err;
+	}
+
+	err = uart_config_get(config->uart, &dt_uart_config);
+	if (err != 0) {
+		LOG_ERR("could not get uart config. err=%d", err);
+		return err;
+	}
+
+	const struct modem_iface_uart_config modem_uart_config = {
+		.rx_rb_buf = data->iface_rb_buf,
+		.rx_rb_buf_len = RYLR_MAX_RESPONSE,
+		.dev = config->uart,
+		.hw_flow_control = dt_uart_config.flow_ctrl,
+	};
+	err = modem_iface_uart_init(&data->mctx.iface, &data->iface_data, &modem_uart_config);
+	if (err != 0) {
+		LOG_ERR("error uart iface init %d", err);
+		return err;
+	}
+
+	data->mctx.driver_data = data;
+	err = modem_context_register(&data->mctx);
+	if (err != 0) {
+		LOG_ERR("error context register %d", err);
+		return err;
+	}
+
+	k_thread_create(&data->modem_rx_thread, data->modem_rx_stack, CONFIG_RYLRXX_RX_STACK_SIZE,
+			modem_rx, &data->mctx, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
+
+	memcpy(data->cmd_buffer, RYLR_CMD_TEST, RYLR_CMD_TEST_LEN_WITH_TERMINATOR);
+	err = rylr_send_cmd_buffer(dev);
+	if (err != 0) {
+		LOG_ERR("could not send AT cmd: %d", err);
+		return err;
+	}
+	LOG_INF("successfully initialized rylr");
+	return err;
+}
+
+static const struct lora_driver_api rylr_lora_api = {
+	.config = rylr_config,
+	.send = rylr_send,
+	.send_async = rylr_send_async,
+	.recv = rylr_recv,
+	.recv_async = rylr_recv_async,
+	.test_cw = rylr_test_cw,
+};
+
+#define RYLR_DEVICE_INIT(n)                                                                        \
+	NET_BUF_POOL_DEFINE(rylr_recv_pool_##n, CONFIG_LORA_RYLRXX_RX_POOL_COUNT,                  \
+			    RYLR_MAX_RESPONSE, 0, NULL);                                           \
+	static K_KERNEL_STACK_DEFINE(modem_rx_stack_##n, CONFIG_RYLRXX_RX_STACK_SIZE);             \
+	K_MSGQ_DEFINE(recv_msgq_##n, sizeof(struct rylr_recv_msg),                                 \
+		      CONFIG_RYLRXXX_UNSOLICITED_RX_MSGQ_SIZE, sizeof(uint32_t));                  \
+	K_SEM_DEFINE(rylr_resp_sem_##n, 0, 1);                                                     \
+	K_SEM_DEFINE(rylr_operation_sem_##n, 1, 1);                                                \
+	static struct rylr_data dev_data_##n = {                                                   \
```

Comment:
Unless you have a good reason the `data` struct shouldn't be initialised to anything, it should be setup in the `init` function. If you are trying to point to other variables that need to be defined, those pointers should be part of the config struct.

Reply:
For variables in the data struct yes. The problem with static init of RAM variables is that it then requires a copy of that entire struct in ROM as well to initialise from. Given the large size of the data struct in this instance, that's a lot of overhead.